etcdserver/quota.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/storage/backend"
+)
+
+type Quota interface {
+	Available(interface{}) bool
+	Cost(interface{}) int
+	Remaining() int64
+}
+
+type backendQuota struct {
+	s               *EtcdServer
+	maxBackendBytes int64
+}
+
+const (
+	leaseOverhead = 32
+	kvOverhead    = 256
+)
+
+func NewBackendQuota(s *EtcdServer) Quota {
+	return &backendQuota{s, backend.InitialMmapSize}
+}
+
+func (b *backendQuota) Available(v interface{}) bool {
+	return b.s.Backend().Size()+int64(b.Cost(v)) < b.maxBackendBytes
```

Comment:
add a `todo: maybe optimize backend.Size()`? i feel this might cost a lot, but i do not have data to support either.